crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs

```diff
@@ -0,0 +1,335 @@
+//! Shared code for [`use_pep695_type_alias`] (UP040) and [`use_pep695_type_parameter`] (UP046)
+
+use std::fmt::Display;
+
+use itertools::Itertools;
+use ruff_python_ast::{
+    self as ast,
+    name::Name,
+    visitor::{self, Visitor},
+    Expr, ExprCall, ExprName, ExprSubscript, Identifier, Stmt, StmtAssign, TypeParam,
+    TypeParamParamSpec, TypeParamTypeVar, TypeParamTypeVarTuple,
+};
+use ruff_python_semantic::SemanticModel;
+use ruff_text_size::{Ranged, TextRange};
+
+pub(crate) use use_pep695_generic_class::*;
+pub(crate) use use_pep695_generic_function::*;
+pub(crate) use use_pep695_type_alias::*;
+
+use crate::checkers::ast::Checker;
+
+mod use_pep695_generic_class;
+mod use_pep695_generic_function;
+mod use_pep695_type_alias;
+
+#[derive(Debug)]
+enum TypeVarRestriction<'a> {
+    /// A type variable with a bound, e.g., `TypeVar("T", bound=int)`.
+    Bound(&'a Expr),
+    /// A type variable with constraints, e.g., `TypeVar("T", int, str)`.
+    Constraint(Vec<Expr>),
```

Comment:
Is it worth using a `std::borrow::Cow` here, so that we can avoid cloning in the common non-`AnyStr` case?

```suggestion
    /// A type variable with constraints, e.g., `TypeVar("T", int, str)`.
    Constraint(Vec<Cow<'a, Expr>>),
```